### PR TITLE
[KeymanWeb] Deadkeys in Stores

### PR DESCRIPTION
**Pull request by jahorton**
This PR extends #613 and allows KMW to support deadkeys in stores, bringing the deadkey feature ever closer to parity with Keyman Desktop.

Checklist:
- [x] Adjust the format of stores to allow for an 'exploded' definition when deadkeys are included.
- [x] Include a cache to auto-store compact stores' exploded forms for reuse.
- [x] Convert `any` and `index` to be compatible with the new store format.
- [x] Extend test cases accordingly.
- [x] Extend the hand-converted reference keyboard accordingly.

### CODE REVIEW

File: web/unit_tests/cases/engine.js [hunk 1]
```diff
@@ -38,6 +38,20 @@ function runEngineRuleSet(ruleSet, defaultNoun) {
   }
 }
 
+var toSupplementaryPairString = function(code){
```
Comment (darcywong00) [inline web/unit_tests/cases/engine.js]: Could you use source/kmwstring.ts for handling SMP?

File: web/source/kmwcallback.ts [hunk 2]
```diff
@@ -425,29 +426,36 @@ class KeyboardInterface {
...
+        if(deadSpec['d'] !== context[i]) {
           mismatch = true;
         } else {
           deadContext[i].set();
         }
       } else if(rule[i]['a'] !== undefined) {
         var anySpec = rule[i] as ContextAny;
         // TODO:  Remove the `string` requirement.
-        if(!this.any(i, context[i] as string, anySpec.a)) {
+        var lookup = (typeof(context[i]) == 'string' ? context[i] as string : {'d': context[i] as number});
```
Comment (darcywong00) [inline web/source/kmwcallback.ts]: Is this TODO resolved now that it handles numbers too?

File: web/source/kmwcallback.ts [hunk 3]
```diff
@@ -628,23 +659,112 @@ class KeyboardInterface {
...
+   * @param       {number}      dn      number of context entries to overwrite
+   * @param       {Object}      Pelem   element to output to 
+   * @param       {string}      s       string to output   
+   * Description  Keyboard output
+   */
+  deleteContext(dn: number, Pelem): void {
+    var context: CachedExEntry;
+
+    // We want to control exactly which deadkeys get removed.
+    if(dn > 0) {
```
Comment (darcywong00) [inline web/source/kmwcallback.ts]: Do you want to range-check the maximum number of context entries allowed to overwrite?

Comment (mcdurdin) [reply web/source/kmwcallback.ts]: I don't think we need to do that here. It doesn't add any safety and is unlikely to arise.

File: web/source/kmwkeyboards.ts [hunk 4]
```diff
@@ -1230,6 +1238,13 @@ class KeyboardManager {
       this.deferredKR.push(Pk);
       return;
     }
+
+    if(Pk['_kmw']) {
+      console.error("The keyboard _kmw property is a reserved field for engine use only; this keyboard is invalid.");
```
Comment (mcdurdin) [inline web/source/kmwkeyboards.ts]: Can we log the id of the keyboard as 'this keyboard' may be ambiguous when loading multiple keyboards...

File: web/unit_tests/cases/engine.js (hunk 1 above)
Comment (mcdurdin) [reply web/unit_tests/cases/engine.js]: I don't think these belong in kmwstring.ts, which is specifically setup as extensions to the `string` type (whether or not extending a built-in type is sensible is a separate discussion). They seem fine here.

**Pull request merged**